src/uu/cp/src/copydir.rs

```diff
@@ -101,14 +101,29 @@ struct Context<'a> {
 }
 
 impl<'a> Context<'a> {
-    fn new(root: &'a Path, target: &'a Path) -> std::io::Result<Self> {
+    /// Creates a new `Context` for copying a directory.
+    ///
+    /// # Parameters
+    ///
+    /// - `root`: The source path from which the directory will be copied.
+    /// - `target`: The target path to which the directory will be copied.
+    /// - `no_target_dir`: A boolean indicating whether the root parent
+    ///    should consider wether or not the target directory exists.
+    ///
+    /// `$ cp -r source target/` Typically behave differently wether `target`
+    /// exists or not, end in `/` or `.`. the `--no-target-dir` flag can be
+    /// used to force the behavior of the command in case where `target` exists,
+    /// and is reflected in the `no_target_dir` parameter.
+    ///
+    fn new(root: &'a Path, target: &'a Path, no_target_dir: bool) -> std::io::Result<Self> {
         let current_dir = env::current_dir()?;
         let root_path = current_dir.join(root);
-        let root_parent = if target.exists() && !root.to_str().unwrap().ends_with("/.") {
-            root_path.parent().map(|p| p.to_path_buf())
-        } else {
-            Some(root_path)
-        };
+        let root_parent =
+            if (target.exists() || no_target_dir) && !root.to_str().unwrap().ends_with("/.") {
```

Comment:
I'm not confident bout this or 

```suggestion
            if (target.exists() && !root.to_str().unwrap().ends_with("/.")) || no_target_dir {
```